CONTRIBUTING.md

```diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+the various PRs over the last few years.
+
+
+#### Errors
+
+Return as much context as possible with errors e.g., if an error was encountered parsing a string
+include the string in the returned error type. If a function consumes costly-to-compute input
+(allocations are also considered costly) it should return the input back in the error type.
+
+More specifically errors should
+
+- be `non_exhaustive`
+- have private fields unless you're very confident they won't change.
+- only derive  `Debug`
+- implement Display using `write_err!()` macro if a variant contains an inner error source.
+- have `Error` suffix
+- implement `std::error::Error` if they are public (feature gated on "std").
+- commit to Send and Sync use `check_pub_error!` macro
+
+```rust
+/// Documentation for the `Error` type.
+#[derive(Debug)]
+#[non_exhaustive]  // Add liberally; if the error type may ever have new variants added.
+pub enum Error {
+    /// Documentation for variant A.
+    A,
+    /// Documention for variant B.
+    B,
+}
+check_pub_error!(Error)
+```
+
+
+#### rustdocs
+
+Be liberal with references to BIPs or other documentation; the aim is that devs can learn about
+Bitcoin by hacking on this codebase as opposed to having to learn about Bitcoin first and then start
+hacking on this codebase. Consider this format
+
+
+```rust
+pub struct BlockHeader {
+    /// Originally protocol version, but repurposed for soft-fork signaling.
+    ///
+    /// ### Relevant BIPs
+    ///
+    /// * [BIP 9 - Version bits with timeout and delay](https://github.com/bitcoin/bips/blob/master/bip-0009.mediawiki) (current usage)
+    /// * [BIP 34 - Block v2, Height in Coinbase](https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki)
+    pub version: i32,
+```
+
+Do use rustdoc subheadings. Do put an empty newline below each heading e.g.,
+
+```rust
+    /// Returns an iterator for the funding UTXOs of the psbt
+    ///
+    /// For each PSBT input that contains UTXO information `Ok` is returned containing that information.
+    /// The order of returned items is same as the order of inputs.
+    ///
+    /// # Errors
+    ///
+    /// The function returns error when UTXO information is not present or is invalid.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if the length of transaction inputs is not equal to the length of PSBT inputs.
+    pub fn iter_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
+```
+
+Add `Panics` section if the function may panic under any circumstance (including array index out of bounds)
+
+#### Attributes
+
+- For feature gated code we use a custom configuration conditional `docsrs` to configure the docs
+  build. This means feature gated code should look something like:
+
+```rust
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for ParseAmountError {
+```
+
+- `#[track_caller]`: Used any time a function may panic and the stack trace is uninformative.
```

Comment:
More like "used on functions that panic on invalid arguments"

Reply:
Cool, done.